mesh/nurbs.cpp

```diff
@@ -341,6 +343,101 @@ void KnotVector::CalcDnShape(Vector &gradn, int n, int i, double xi) const
 
 }
 
+void KnotVector::FindMaxima(Array<int> &ks,
+                            Vector &xi,
+                            Vector &u)
+{
+   Vector shape(Order+1);
+   Vector maxima(GetNCP());
+   double arg1, arg2, arg, max1, max2, max;
+
+   xi.SetSize(GetNCP());
+   u.SetSize(GetNCP());
+   ks.SetSize(GetNCP());
+   for (int j = 0; j <GetNCP(); j++)
+   {
+      maxima[j] = 0;
+      for (int d = 0; d < Order+1; d++)
+      {
+         int i = j - d;
+         if (isElement(i))
+         {
+            arg1 = 1e-16;
+            CalcShape(shape, i, arg1);
+            max1 = shape[d];
+
+            arg2 = 1-(1e-16);
+            CalcShape(shape, i, arg2);
+            max2 = shape[d];
+
+            arg = (arg1 + arg2)/2;
+            CalcShape(shape, i, arg);
+            max = shape[d];
+
+            while ( ( max > max1 ) || (max > max2) )
+            {
+               if (max1 < max2)
+               {
+                  max1 = max;
+                  arg1 = arg;
+               }
+               else
+               {
+                  max2 = max;
+                  arg2 = arg;
+               }
+
+               arg = (arg1 + arg2)/2;
+               CalcShape ( shape, i, arg);
+               max = shape[d];
+            }
+
+            if (max > maxima[j])
+            {
+               maxima[j] = max;
+               ks[j] = i;
+               xi[j] = arg;
+               u[j]  = getKnotLocation(arg, i+Order);
+            }
+         }
+      }
+   }
+}
+
+// Routine from "The NURBS book" - 2nd ed - Piegl and Tiller
+// Algorithm A9.1 p. 369
+void KnotVector::FindInterpolant(Array<Vector*> &x)
+{
+   int order = GetOrder();
+   int ncp = GetNCP();
+
+   // Find interpolation points
```

Comment:
Here, the knotvector coordinates are chosen as the maxima of the basis functions. Once the knotvector coordinates are chosen, a system is solved for each control point. For Algorithm A9.1, doesn't the NURBS book discuss other ways to choose the knotvector coordinates, which are simpler and easier to compute than `FindMaxima`? For example, can't the coordinates be choosen as relative chord lengths? Have you compared `FindMaxima` with other options?

Reply:
Good question! I have not compared `FindMaxima` with other methods. From what I understand from the NURBS book there is not necessary a right or wrong answer, but of course this book can be a bit outdated.
We choose for `FindMaxima` for two reasons:
- The spacing between the controlpoints is directly dependend on the spacing between the knots in the `KnotVector`
- We have the idea that inverting the matrix at line 433 of nurbs.cpp in `FindInterpolant` is easier when evaluating the shapefunctions at their maxima. 

You can actually apply this method in combination with chordwise spacing. I have the plan to add a C-mesh around a hydrofoil in another pull request where I apply that. However, I am currently still working on that in the branch [nurbs-auto-knot2edge-dev](https://github.com/mfem/mfem/tree/nurbs-auto-knot2edge-dev).

Reply:
The rows of the matrix add up to 1 due to partition of unity property of the basis. If we choose the collocation points to be the maxima of the shapefunctions, we make the matrix as diagonally dominant as possible (assuming the correct ordering of the points). This seemed (based on my expert intuition :) ) to be favorable in terms conditioning (and proofs?).